### PR TITLE
feat(json-n-jsw): support typings for jsw and json files

### PR DESCRIPTION
**Pull request by matanlb**

### CODE REVIEW

File: configs/tsconfig.base.json [hunk 1]
```diff
@@ -2,6 +2,8 @@
   "compilerOptions": {
     "lib": ["ES2020"],
     "allowJs": true,
+    "moduleResolution": "node",
```
Comment (nemni8) [inline configs/tsconfig.base.json:5]: why do we need here `"moduleResolution": "node"` ?

File: package.json [hunk 2]
```diff
@@ -3,7 +3,7 @@
   "version": "0.1.108",
   "description": "Type definitions for Corvid by Wix",
   "scripts": {
-    "lint": "eslint .",
+    "lint": "eslint ./src/** ./test/** ./scripts/**",
```
Comment (nemni8) [inline package.json:6]: consider separating into a different commit

File: src/dynamicTypes/widget.js [hunk 3]
```diff
@@ -18,10 +18,19 @@ const getMembersTypes = members =>
...
-const getWidgetTypeDeclarations = ({ name, members = {}, events = {} }) => {
+const getWidgetDeclaration = ({ name, members = {}, events = {} }) => {
   return `interface ${name} extends $w.IFrame { 
     ${getMembersTypes(members)} ${getEventHandlersTypes(events)}}`;
 };
+
 module.exports = {
-  getWidgetTypeDeclarations
+  getWidgetDeclaration,
+  getWidgetsTypeDeclarations: function getWidgetsTypeDeclarations(widgets) {
```
Comment (nemni8) [inline src/dynamicTypes/widget.js:28]: consider declaring the function above (imo more readable)

File: src/index.js [hunk 4]
```diff
@@ -1,47 +1,23 @@
 const fullCorvidTypes = require("../fullCorvidTypes.json");
 const wixModulesNames = require("../wixModules.json");
 const { TS_CONFIG_PATHS } = require("./constants");
-const { getWidgetTypeDeclarations } = require("./widget");
```
Comment (nemni8) [inline src/index.js:4]: nice!!

File: src_types/backend/index.d.ts [hunk 5]
```diff
@@ -1,3 +1,2 @@
 /// <reference path="../common/declaration.d.ts" />
-/// <reference path="../common/void$w$w.d.ts" />
-/// <reference path="../common/jsonFiles.d.ts" />
\ No newline at end of file
+/// <reference path="../common/void$w$w.d.ts" />
```
Comment (nemni8) [inline src_types/backend/index.d.ts:2]: consider separating the revert commit

File: src/index.js [hunk 6]
```diff
@@ -54,30 +30,35 @@ module.exports = {
...
+        ...getJsonTypeDeclarations(),
         ...getNpmDependenciesTypesDeclarations(dependencies)
       ];
     }
   },
   getWixModulesList: () => wixModulesNames,
-  getWidgetTypeDeclarations,
-  getPageElementsTypeDeclarations,
-  getNpmDependenciesTypesDeclarations
+  getWidgetTypeDeclarations: getWidgetDeclaration,
```
Comment (nemni8) [inline src/index.js:60]: 👍

File: test/it/code-samples/positive/public/importJson.js [hunk 7]
```diff
@@ -1 +1,3 @@
-import data from 'someFiles.json'
```
Comment (nemni8) [inline test/it/code-samples/positive/public/importJson.js:1]: consider splitting/renaming this code to a dedicated test (public-import-json) as you did in pages and backend.

File: test/utils/dynamicTypings.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
+const _ = require("lodash");
+
+const {
+  getWidgetTypeDeclarations,
+  getPageElementsTypeDeclarations,
+  getAmbientModuleDeclaration,
+  getJswTypeDeclaration
+} = require("../../dist/corvidTypes.umd");
+
+module.exports = function getDynamicTypings({
```
Comment (nemni8) [inline test/utils/dynamicTypings.js:10]: why do we need this redundant `function getDynamicTypings`?

File: configs/tsconfig.base.json (hunk 1 above)
Comment (matanlb) [reply configs/tsconfig.base.json:5]: Cause `resolveJsonModule` doesn't work without it..

**Comment by nemni8** [comment]
good work! 
I think you should split your changes into smaller commits (revert old JSON solution, eslint rule, implement JSON, implement jsw, test).
please wait with the merge, I know that @dor-itzhaki wanted to go over it before you merge this code.

File: package.json (hunk 2 above)
Comment (matanlb) [reply package.json:6]: Way so?
The lint command was busted, I fixed it cause I needed it

File: src/dynamicTypes/widget.js (hunk 3 above)
Comment (matanlb) [reply src/dynamicTypes/widget.js:28]: True, done

File: src/index.js (hunk 4 above)
Comment (matanlb) [reply src/index.js:4]: :)

File: src_types/backend/index.d.ts (hunk 5 above)
Comment (matanlb) [reply src_types/backend/index.d.ts:2]: That's strange, I did a revert commit!

Comment (matanlb) [reply src_types/backend/index.d.ts:2]: Ohh ok I reverted the removal of that cause I was going to change the implementation

File: test/utils/dynamicTypings.js (hunk 8 above)
Comment (matanlb) [reply test/utils/dynamicTypings.js:10]: Cause we had the exact same code between positive and negative checks, and it was getting more complex. so instead of changing both places I extracted this function

File: test/it/code-samples/positive/public/importJson.js (hunk 7 above)
Comment (matanlb) [reply test/it/code-samples/positive/public/importJson.js:1]: Done

File: src_types/backend/index.d.ts (hunk 5 above)
Comment (matanlb) [reply src_types/backend/index.d.ts:2]: So, story is..
I removed this code since I was implementing this in code-code.
Then we decided to stay away from regex's more so I decided to move it BACK to to corvid-types, so I reverted my deletion of this, and this PR changes the implementation of it.

**Comment by matanlb** [comment]
> good work!
> I think you should split your changes into smaller commits (revert old JSON solution, eslint rule, implement JSON, implement jsw, test).
> please wait with the merge, I know that @dor-itzhaki wanted to go over it before you merge this code.

Thanks :)
I'll wait for @dor-itzhaki review.
Regarding splitting the PR. That might have been a good idea before I started working on this, but since I've been working on both the JSON and JSW for a while now I see how similar they are so inlight of the amount of changes needed does't make much since to me to split them.
As far as the eslint go's, that's just a happy accident I had to fix it :)

File: configs/tsconfig.base.json (hunk 1 above)
Comment (dor-itzhaki) [reply configs/tsconfig.base.json:5]: what else does it affect?

File: package.json (hunk 2 above)
Comment (dor-itzhaki) [reply package.json:6]: what wasn't working before ?

File: src/index.js [hunk 9]
```diff
@@ -40,6 +42,7 @@ module.exports = {
         ...fullCorvidTypes.BASE,
         ...fullCorvidTypes.BACKEND,
         ...getJswTypeDeclarations(),
+        ...getJsonTypeDeclarations(),
         ...getNpmDependenciesTypesDeclarations(dependencies)
```
Comment (dor-itzhaki) [inline src/index.js]: why aren't these 2 part of the base or some other common thingy ?

**Review by dor-itzhaki — COMMENTED** [review]
can you please split the refactor of the dynamic typings from the jsw solution ?
i'm losing myself and can't really review it this way...

**Comment by matanlb** [comment]
> can you please split the refactor of the dynamic typings from the jsw solution ?
> i'm losing myself and can't really review it this way...

@dor-itzhaki I can but only downstream from this branch not off master

File: src/index.js [hunk 10]
```diff
@@ -1,47 +1,18 @@
...
+  getNpmDependenciesTypesDeclarations
+} = require("./dynamicTypes/npmDependencies");
+const {
+  getPageElementsTypeDeclarations,
+  getPageElementsRawDeclarations
+} = require("./dynamicTypes/elementsMap");
+const {
+  getWidgetsTypeDeclarations,
+  getWidgetDeclaration
+} = require("./dynamicTypes/widget");
```
Comment (dor-itzhaki) [inline src/index.js:15]: can we make all dynamicTypes expose the same api (getTypeDeclaration/getWhateverDeclaration) so that we can just import all via some index and map through them with the same function names ?

File: src/index.js [hunk 11]
```diff
@@ -77,7 +46,7 @@ module.exports = {
     }
   },
   getWixModulesList: () => wixModulesNames,
-  getWidgetTypeDeclarations,
-  getPageElementsTypeDeclarations,
-  getNpmDependenciesTypesDeclarations
+  getWidgetTypeDeclarations: getWidgetDeclaration,
+  getPageElementsTypeDeclarations: getPageElementsRawDeclarations,
+  getAmbientModuleDeclaration
```
Comment (dor-itzhaki) [inline src/index.js]: why are these methods exposed?

File: test/it/code-samples/positive/backend-import-jsw/core.js [hunk 12]
```diff
@@ -0,0 +1,4 @@
+import main, { something } from 'backend/file.jsw'
```
Comment (dor-itzhaki) [inline test/it/code-samples/positive/backend-import-jsw/core.js:1]: this file and others below belong to a different commit

File: test/it/negative.spec.js [hunk 13]
```diff
@@ -1,73 +0,0 @@
-const compiler = require("../utils/compiler");
```
Comment (dor-itzhaki) [inline test/it/negative.spec.js]: i guess you deleted the test files by accident ?

File: src/dynamicTypes/jswFiles.js [hunk 14]
```diff
@@ -0,0 +1,18 @@
+// TODO: Get the project file tree as a paramter and generate a d.ts containing a deceleration
+//       foreach .jsw file (i.e remove the wildcard)
+
+const getJswRawDeclaration = () => "declare module 'backend/*.jsw';";
+
+function getJswTypeDeclarations() {
+  return [
+    {
+      path: "/patches/jswModules.d.ts",
```
Comment (dor-itzhaki) [inline src/dynamicTypes/jswFiles.js:9]: where do these paths go to ?

File: src/index.js [hunk 15]
```diff
@@ -34,19 +39,22 @@ module.exports = {
...
         ...getNpmDependenciesTypesDeclarations(dependencies)
       ];
     }
   },
   getWixModulesList: () => wixModulesNames,
   getWidgetTypeDeclarations: getWidgetDeclaration,
   getPageElementsTypeDeclarations: getPageElementsRawDeclarations,
-  getAmbientModuleDeclaration
+  getAmbientModuleDeclaration,
+  getJswTypeDeclaration: getJswRawDeclaration
```
Comment (dor-itzhaki) [inline src/index.js:63]: why do we need to expose this function ? it should be internal

File: src/index.js [hunk 16]
```diff
@@ -25,6 +29,7 @@ module.exports = {
       return [
         ...fullCorvidTypes.BASE,
         ...fullCorvidTypes.PAGES,
+        ...getJswTypeDeclarations(),
```
Comment (dor-itzhaki) [inline src/index.js:33]: should we unite common thing like jsw, dependencies, other ? under the base context or some other common function ?

File: test/it/negative.spec.js [hunk 17]
```diff
@@ -3,12 +3,8 @@ const listSubDirectories = require("../utils/listSubDirectories");
...
 const testParser = require("../utils/testParser");
 
-const {
-  declarations,
-  getWidgetTypeDeclarations,
-  getPageElementsTypeDeclarations,
-  getNpmDependenciesTypesDeclarations
-} = require("../../dist/corvidTypes.umd.js");
+const getDynamicTypings = require("../utils/dynamicTypings");
+const { declarations } = require("../../dist/corvidTypes.umd.js");
```
Comment (dor-itzhaki) [inline test/it/negative.spec.js:7]: @nemni8 are we using the "umd" output or should we actually test the "esm" one ?

File: test/utils/writeTypingsHelper.js [hunk 18]
```diff
@@ -26,7 +26,11 @@ module.exports = {
     });
     return dir;
   },
-  dynamic: (rootTestPath, context, { elementsMap, widgets, dependencies }) => {
+  dynamic: (
+    rootTestPath,
+    context,
+    { elementsMap, widgets, dependencies, jswFiles }
```
Comment (dor-itzhaki) [inline test/utils/writeTypingsHelper.js:31]: should we just get a list of {name, content} objects and write them to files without knowing their categorization?

**Review by dor-itzhaki — COMMENTED** [review]
the 1st & 2nd commits are still mixed... i still reviewed the best i could.
in general it looks good, but i've added different comments/questions...

File: src/index.js (hunk 15 above)
Comment (dor-itzhaki) [reply src/index.js:63]: add comment/function to explain that this is used by the cli.
maybe don't add the jsw until it is used in the cli.
(we need to organize the cli's api)

File: package.json (hunk 2 above)
Comment (matanlb) [reply package.json:6]: Nope 🤷‍♂️

File: src/index.js (hunk 9 above)
Comment (matanlb) [reply src/index.js]: done

File: package.json (hunk 2 above)
Comment (dor-itzhaki) [reply package.json:6]: please add the lint script to executed as part of the test script as well:
`"test": "yarn lint && jest"`

File: src/index.js (hunk 10 above)
Comment (dor-itzhaki) [reply src/index.js:15]: for me, it would be much more understandable and organized.
then you can just import the full object and call the relevant function

File: src/index.js (hunk 15 above)
Comment (matanlb) [reply src/index.js:63]: Done

File: test/it/negative.spec.js (hunk 13 above)
Comment (matanlb) [reply test/it/negative.spec.js]: IDK man it's still there... maybe it was a mistake made in the commit split somehow.
Final result is that it is there

File: src/index.js (hunk 16 above)
Comment (matanlb) [reply src/index.js:33]: United jsw & json as commons.
Made a unified API for all and imported them via index file

File: src/index.js (hunk 15 above)
Comment (matanlb) [reply src/index.js:63]: Added a comment but we still need these exports for the tests.
We run every test for the editor (in which we get the typings via the page/public/backend method) and another as if we are using the CLI (in which we use the exported functions to generate the typings and see that they work)

File: test/utils/writeTypingsHelper.js (hunk 18 above)
Comment (matanlb) [reply test/utils/writeTypingsHelper.js:31]: Done 👍
Turned out real nice I think :)

File: src/dynamicTypes/index.js [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+const {
+  getWidgetRawDeclaration,
+  getWidgetsTypeDeclarations
+} = require("./widget");
+const { getJswRawDeclaration, getJswTypeDeclarations } = require("./jswFiles");
+const getJsonTypeDeclarations = require("./jsonFiles");
+
+module.exports = {
+  packages: {
+    getRaw: getPackageRawDeclaration,
```
Comment (dor-itzhaki) [inline src/dynamicTypes/index.js:18]: you can go one step further and make each file export getRaw & getFiles, so you don't need to split them here
```
const npmDependencies = require("./npmDependencies")
...
module.exports = {npmDependencies, ...}
```

File: test/utils/writeTypingsHelper.js [hunk 20]
```diff
@@ -26,21 +26,19 @@ module.exports = {
     });
     return dir;
   },
-  dynamic: (rootTestPath, context, { elementsMap, widgets, dependencies }) => {
+  dynamic: (rootTestPath, context, dynamicTypingsFiles) => {
```
Comment (dor-itzhaki) [inline test/utils/writeTypingsHelper.js:29]: i would still make "dynamicTypingsFiles" an array of ({content, fileName}), where content is always a string (not an array).
no need to know all of that here

File: src/dynamicTypes/index.js (hunk 19 above)
Comment (matanlb) [reply src/dynamicTypes/index.js:18]: Nice, done

File: test/utils/writeTypingsHelper.js (hunk 20 above)
Comment (matanlb) [reply test/utils/writeTypingsHelper.js:29]: Yeah but that is how the data is store and processed.
We can a bit of that in the positive/negative.spec files that uses this method but then we'd be writing the same code twice which is what this function was meant to solve.

Comment (matanlb) [reply test/utils/writeTypingsHelper.js:29]: So I'd rather just leave it like that.
This way we can also introduce new typings concepts with ease, just add the key

File: package.json (hunk 2 above)
Comment (matanlb) [reply package.json:6]: Nice, done

File: test/utils/writeTypingsHelper.js (hunk 20 above)
Comment (dor-itzhaki) [reply test/utils/writeTypingsHelper.js:29]: i'm not sure i follow.... why doesn't `getDynamicTypings` just flatten its return value ?
